### DISCUSSION TITLE
Async support w/ httpx

### DISCUSSION THREAD

**Discussion by TheOnlyWayUp** [sigma67/ytmusicapi#545] [Ideas]
Async support would be a good idea. The `httpx` package is a popular choice for API Wrappers that support both.

I'm interested in contributing, I'll make a PR. 

@sigma67 Could we talk further on Discord? I'm `TheOnlyWayUp` on there.

**Comment by sigma67** [comment]
Feel free to make a fork, I'm not interested in supporting async in the main lib

https://github.com/sigma67/ytmusicapi/discussions/541

However, I would be open to restructuring the project to support both a sync and async version on top of a core project that takes care of the domain-specific parts such as parsing and models (`ytmusicapi-core`)

**Reply by sigma67** [reply]
My personal issue with `httpx` is that it's in beta state. I really don't want to swap out a battle-tested and PSF-supported library like `requests` for a beta project that may not be supported in a year or two

Especially because swapping out `requests` would not be a small effort at this point, while also introducing very different dependencies through httpx

**Reply by TheOnlyWayUp** [reply]
> Especially because swapping out `requests` would not be a small effort at this point, while also introducing very different dependencies through httpx

@sigma67 Would `aiohttp` be up to the mark?

>  domain-specific parts such as parsing and models

I see that the library prefers returning raw JSON. Do you think `pydantic` models could be better?

For example:
```py
search_results = await ytmusic.search("Oasis Wonderwall")
search_results[0].videoId or search_results[0].video_id
```

Along with the type validation, Pydantic models are useful because of the IDE Support they provide:

> ![](https://prisma-client-py.readthedocs.io/en/stable/showcase.gif)
> ~ *https://prisma-client-py.readthedocs.io/en/stable/*

Is an example of the IDE Autocomplete when using Pydantic Models & TypedDicts. It's a better user experience, and for backwards compatibility, the `.raw` attribute would be provided.

**Reply by sigma67** [reply]
pydantic support has been planned for a long time, see https://github.com/sigma67/ytmusicapi/issues/307

I just need to find some free time to get started, maybe next month.

`aiohttp` doesn't support sync requests as far as I know? Either way, need to tackle one topic at a time. Pydantic definitely has priority

**Reply by sigma67** [reply]
Contributions are very welcome towards Pydantic migration!

**Reply by TheOnlyWayUp** [reply]
Awesome! Let's get the core module & parsing done, sync/async will be just an abstraction over it.

**Reply by TheOnlyWayUp** [reply]
@sigma67 Hi! I'm still interested in contributing, things got hectic. I'll start contributing from the 20th.

**Reply by sigma67** [reply]
Sure, I think it's good timing now that most of the podcast stuff is finished

I'll be busy in the last week of the month, so feedback may take a bit, but I'll make sure to get back to it

**Reply by TheOnlyWayUp** [reply]
I'm working on it.
1. Create Pydantic BaseModels
2. Functions output to BaseModels
3. Update tests for the same
---
4. Look into request logic and see how requests and parsing are separated 
5. Add aiohttp (this can wait)

**Reply by TheOnlyWayUp** [reply]
yo @sigma67, from a typing perspective, it's better to have `yt.get_search_suggestions('...', detailed_runs=True)` be a different function. `detailed_runs=False` returns `list[str]`, truthy returns `list[dict]`.

**Reply by sigma67** [reply]
Yeah I didn't write that function. Not a big fan of changing the return type based on a function parameter as well (although you could probably do it using typing overloads as I did with nav, but that's beside the point)

We're doing 2.0 with these changes so some breaking changes are ok. I don't think this function is used by that many - take that with a grain of salt

**Reply by sigma67** [reply]
Hi @TheOnlyWayUp , I saw you made some changes in your fork, care to submit a PR? :)

**Reply by TheOnlyWayUp** [reply]
@sigma67 I'll be done by Friday!

**Reply by TheOnlyWayUp** [reply]
I'm really sorry for flaking out on you, hope someone else can contribute to this
